hw/ip/pattgen/doc/_index.md

```diff
@@ -0,0 +1,143 @@
+1. Program the desired number of pattern repetitions using one of the fields {{< regref "PATT_SIZE.REPS_CH0" >}} for Channel 0, or {{< regref "PATT_SIZE.REPS_CH1" >}} for Channel 1.
+Note that since the allowed number of pattern repetitions ranges from 1-1024, the value of this field should be one less than the desired repetition count.
+For example, to repeat a pattern 30 times on Channel 1, a value of 29 should written to the field {{< regref "PATT_SIZE.REPS_CH1" >}}. 
+1. Finally to start the pattern, set the {{< regref "PATT_CTRL.ENABLE_CH0" >}} to enable Channel 0 or set {{< regref "PATT_CTRL.ENABLE_CH1" >}} to enable Channel 1.
+Set both bits simultaneously, to start both channel patterns at the same time.
+
+The pattern can be halted on either channel by setting the corresponding `PATT_CTRL.ENABLE` bit to 0 for the desired channel.
+Note that re-enabling either channels restarts the pattern sequence for that channel from the beginning.
+
+PLease note that writes to a channel's configuration registers have no effect while a pattern is running on that channel.
+For operational simplicity, the configuration registers are only transferred into the internal finite state machines while a channel is disabled.
+Changes to the configuration registers only take effect once the channel has been disabled and re-enabled.
+
+## Block Diagram
+
+![](pattgen_block_diagram.svg)
+
+## Hardware Interfaces
+
+{{< hwcfg "hw/ip/pattgen/data/pattgen.hjson" >}}
+
+## Design Details 
+
+The design consists of two identical, and independent finite state machines, each an instance of module `pattgen_fsm`).
+Each FSM is essentially three nested counters, with one counter to control the clock division, another to count out the sequence bits, and a third to keep count of the number of repetitions.
+
+Each FSM consists of.
+- Inputs::
+    - `clk_io`, `reset`, `enable`, `clk_pre_divide`, `pattern`, `pattern_size`, `polarity`,  and `n_repeats`
+- Outputs:
+    - `pda` and `pcl`
+- a single state variable with three states `IDLE`, `ACTIVE`, and `END`,.
+- a clock-divide counter, `clk_div`. a single clock-divide flop, `clk_int`, and
+- two two internal counters `bit_ctr` and `repeat_ctr`
+
+Each FSM is disabled when `enable` is low.
+Disabling the FSM is equivalent to reset, and both operations place the FSM in the `IDLE` state.
+While in `IDLE`, the other state machine registers assume their default states:
+The internal counters, the clock-divide, `bit_ctr` and `repeat_ctr` all reset to 0, as does `clk_int`.
+
+Once the FSM is enabled, it transitions to the `ACTIVE` state.  The clock-divide counter `clk_div` increments every cycle, except when it overflows at `clk_pre_divide`.
+Then it resets to 0, toggling `clk_int` in the process.
+The results in an internal clock frequency of:
+$$f_\textsf{clk\_int}=\frac{f_\textsf{clk\_io}}{2\times\textsf{clkPreDivide}}\textrm{ for }x\in{0, 1}$$.
```

Comment:
This formula doesn't match the one used in the "theory of operations" text.

Reply:
Thanks for catching that.  This is another change that happened the document was revised.   Will fix in upcoming PR.